org/mozilla/jss/util/Password.java

```diff
@@ -172,9 +177,24 @@ public synchronized Object clone() {
      */
     @Deprecated
     protected void finalize() throws Throwable {
-        if(!cleared) {
-            logger.warn("Password was garbage collected before it was cleared.");
+
+        if (!cleared) {
+
+            StringWriter sw = new StringWriter();
+            PrintWriter out = new PrintWriter(sw);
+            out.println("Uncleared Password object");
```

Comment:
Just one request. Could you put "created at" here, so it is clear that this is where the password was *created* and isn't the stack trace of this finalize method? My 2c. but I think it'll help readability. 

Also, do we want to make this a warning and say something like "please report this" so people know this is something we should deal with and not something they e.g., configured wrong?

Reply:
I agree about the "created at", but I'm not sure how the whole output should look like because each line starts with an "at" already. 

About the "warning", we do have other warnings in debug log or systemd journal which is caused by bugs in our code, so the user is expected to report that to the developers. A warning is basically something that needs a closer inspection, but not necessarily a problem. It doesn't necessarily mean user error.

Perhaps you could provide an example how the whole output should look like?

Reply:
If I were to be picky, I'd say:

```
WARNING: Uncleared Password object created at org.mozilla.jss.crypto.PBEKeyGenParams.<init>(PBEKeyGenParams.java:37)
        from org.mozilla.jss.pkcs12.MacData.<init>(MacData.java:118)
        from org.mozilla.jss.pkcs12.PFX.verifyAuthSafes(PFX.java:156)
        from org.mozilla.jss.netscape.security.pkcs.PKCS12Util.loadFromByteArray(PKCS12Util.java:835)
        from org.mozilla.jss.netscape.security.pkcs.PKCS12Util.loadFromFile(PKCS12Util.java:824)
        from com.netscape.cmstools.pkcs12.PKCS12CertFindCLI.execute(PKCS12CertFindCLI.java:119)
        from org.dogtagpki.cli.CLI.execute(CLI.java:345)
        from org.dogtagpki.cli.CLI.execute(CLI.java:345)
        from org.dogtagpki.cli.CLI.execute(CLI.java:345)
        from com.netscape.cmstools.cli.MainCLI.execute(MainCLI.java:667)
        from com.netscape.cmstools.cli.MainCLI.main(MainCLI.java:703)
```

But perhaps just adding the word created would suffice:

```
WARNING: Uncleared Password object created
        at org.mozilla.jss.crypto.PBEKeyGenParams.<init>(PBEKeyGenParams.java:37)
        at org.mozilla.jss.pkcs12.MacData.<init>(MacData.java:118)
        at org.mozilla.jss.pkcs12.PFX.verifyAuthSafes(PFX.java:156)
        at org.mozilla.jss.netscape.security.pkcs.PKCS12Util.loadFromByteArray(PKCS12Util.java:835)
        at org.mozilla.jss.netscape.security.pkcs.PKCS12Util.loadFromFile(PKCS12Util.java:824)
        at com.netscape.cmstools.pkcs12.PKCS12CertFindCLI.execute(PKCS12CertFindCLI.java:119)
        at org.dogtagpki.cli.CLI.execute(CLI.java:345)
        at org.dogtagpki.cli.CLI.execute(CLI.java:345)
        at org.dogtagpki.cli.CLI.execute(CLI.java:345)
        at com.netscape.cmstools.cli.MainCLI.execute(MainCLI.java:667)
        at com.netscape.cmstools.cli.MainCLI.main(MainCLI.java:703)
```

(I recognize this isn't python so we can't easily slice from the first element).

But the key point is to note that this is where it is *created*, not where the `finalize()` method is being called from. (Does the latter give us any value? I'm not sure).

Reply:
I'd pick the first one, but the last line probably should say something like this instead:
```
Please report this issue at https://pagure.io/dogtagpki/issues.
```
The second option might be confusing if people only read the first line.

The stack trace for the finalize() itself is not very useful:
```
java.lang.Throwable
	at org.mozilla.jss.util.Password.finalize(Password.java:186)
	at java.lang.System$2.invokeFinalize(System.java:1270)
	at java.lang.ref.Finalizer.runFinalizer(Finalizer.java:102)
	at java.lang.ref.Finalizer.access$100(Finalizer.java:34)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:217)
```
Let me update the patch. Thanks!

Reply:
How about:
> Please report the above trace to your software vendors.

JSS is a library that could be used in other software (such as Candlepin) -- reporting this to us would mean we'd need to track down whoever else is using our library.

Reply:
OK, sounds good to me.